File: discipline4/app/src/main/java/com/example/app/vaults/VaultName.kt
package com.example.app

import com.example.app.*
import androidx.room.Entity

/**
 * Represents a vault name with length constraints
 */
@Entity
@JvmInline
public value class VaultName private constructor(val value: String) {
  
  companion object {
    const val MINIMUM_LENGTH = 1
    const val MAXIMUM_LENGTH = 300
    
    /**
     * Creates a VaultName with validation
     */
    fun create(string: String): Tried<VaultName, TextualError> {
      if (string.length < MINIMUM_LENGTH || string.length > MAXIMUM_LENGTH) {
        return Tried.failure(
          TextualError.create("Creating VaultName from string")
            .addMessage("String violates length invariants")
            .addNumberAttachment("Minimum length", MINIMUM_LENGTH.toDouble())
            .addNumberAttachment("Maximum length", MAXIMUM_LENGTH.toDouble())
            .addNumberAttachment("Provided string length", string.length.toDouble())
            .addStringAttachment("Provided string", string)
        )
      }
      
      return Tried.success(VaultName(string))
    }

    fun createOrThrow(string: String): VaultName {
      if (string.length < MINIMUM_LENGTH || string.length > MAXIMUM_LENGTH) {
        throw TextualError.create("Creating VaultName from string")
          .addMessage("String violates length invariants")
          .addNumberAttachment("Minimum length", MINIMUM_LENGTH.toDouble())
          .addNumberAttachment("Maximum length", MAXIMUM_LENGTH.toDouble())
          .addNumberAttachment("Provided string length", string.length.toDouble())
          .addStringAttachment("Provided string", string)
      }

      return VaultName(string)
    }
  }
  
  fun isEqualTo(other: VaultName): Boolean {
    return value == other.value
  }
  
  fun length(): Int {
    return value.length
  }

  override fun toString(): String {
    return "VaultName($value)"
  }
}

File: discipline_ts/core/vault_name.ts
import { TextualError, Unique } from "../x.ts";

const BRAND = Symbol();

export type VaultName = Unique<typeof BRAND, "VaultData", string>;

const MINIMUM_LENGTH = 1;
const MAXIMUM_LENGTH = 300;

const construct = (string: string): VaultName => {
  return string as VaultName;
};

const wrapOrThrow = (string: string): VaultName => {
  if (
    string.length < MINIMUM_LENGTH 
    || 
    string.length > MAXIMUM_LENGTH
  ) {
    const textualError = TextualError.create("Creating VaultName from string");
    TextualError.addMessage(textualError, "String violates length invariants. ");
    TextualError.addNumberAttachment(textualError, "Minimum length", MINIMUM_LENGTH);
    TextualError.addNumberAttachment(textualError, "Maximum length", MAXIMUM_LENGTH);
    TextualError.addNumberAttachment(textualError, "Provided string's length", string.length);
    TextualError.addStringAttachment(textualError, "Provided string", string);
    throw TextualError.toJsError(textualError);
  }

  return construct(string);
};

const getValue = (it: VaultName): string => {
  return it;
};

const isEqualTo = (it: VaultName, rhs: VaultName): boolean => {
  return it === rhs;
};

export const VaultName = {
  MINIMUM_LENGTH,
  MAXIMUM_LENGTH,
  wrapOrThrow,
  construct,
  isEqualTo,
  getValue,
};